packages/editor/src/dataviews/types.ts

```diff
@@ -0,0 +1,27 @@
+type PostStatus =
+	| 'published'
+	| 'draft'
+	| 'pending'
+	| 'private'
+	| 'future'
+	| 'auto-draft'
+	| 'trash';
+
+type ReservedPostTypes = 'template' | 'template-part';
+
+export interface BasePost {
+	status?: PostStatus;
+	title: string | { rendered: string };
+}
+
+export interface TemplateOrTemplatePart extends BasePost {
+	type: 'template' | 'template-part';
+	source: string;
+	has_theme_file: boolean;
+}
+
+export interface OtherPost extends BasePost {
+	type: Exclude< string, ReservedPostTypes >;
+}
```

Comment:
I think we can simplify this section, the following might work:

```suggestion
export interface Post {
	status?: PostStatus;
	title: string | { rendered: string };
	type: string;
}

export interface TemplateOrTemplatePart extends Post {
	type: 'template' | 'template-part';
	source: string;
	has_theme_file: boolean;
}
```

The main reason is that `Exclude< string, ReservedPostTypes >` can't really do anything to narrow the type, it's just `string`. The type "any string except these two literals" is too broad and just becomes `string`. Any `OtherPost` would also satisfy the `TemplateOrTemplatePart` type.

A type guard would be suitable to narrow the type in application code:

```tsx
function isTemplateOrTemplatePartPost( p: Post ): p is TemplateOrTemplatePart {
    return p.type === 'template' || p.type === 'template-part';
}

declare var p: Post;
if ( isTemplateOrTemplatePartPost(p)) {
    p.has_theme_file; // Now `p` is of type `TemplateOrTemplatePart`
}
```